lib/location/loc_core.c

```diff
@@ -0,0 +1,515 @@
+			(char *)loc_method_api_get(current_event_data.method)->method_string,
+			current_event_data.method);
+		/* Logging v1 doesn't support double and float logging. Logging v2 would support
+		 * but that's up to application to configure.
+		 */
+		sprintf(latitude_str, "%.06f", current_event_data.location.latitude);
+		LOG_DBG("  latitude: %s", log_strdup(latitude_str));
+		sprintf(longitude_str, "%.06f", current_event_data.location.longitude);
+		LOG_DBG("  longitude: %s", log_strdup(longitude_str));
+		sprintf(accuracy_str, "%.01f", current_event_data.location.accuracy);
+		LOG_DBG("  accuracy: %s m", log_strdup(accuracy_str));
+		if (current_event_data.location.datetime.valid) {
+			LOG_DBG("  date: %04d-%02d-%02d",
+				current_event_data.location.datetime.year,
+				current_event_data.location.datetime.month,
+				current_event_data.location.datetime.day);
+			LOG_DBG("  time: %02d:%02d:%02d.%03d UTC",
+				current_event_data.location.datetime.hour,
+				current_event_data.location.datetime.minute,
+				current_event_data.location.datetime.second,
+				current_event_data.location.datetime.ms);
+		}
+		LOG_DBG("  Google maps URL: https://maps.google.com/?q=%s,%s",
+			log_strdup(latitude_str), log_strdup(longitude_str));
+	} else {
+		/* Do fallback to next preferred method */
+		previous_loc_method = current_event_data.method;
+		current_loc_method_index++;
+		if (current_loc_method_index < current_loc_config.methods_count) {
+			requested_loc_method =
+				current_loc_config.methods[current_loc_method_index].method;
+
+			LOG_WRN("Failed to acquire location using '%s', "
+				"trying with '%s' next",
+				(char *)loc_method_api_get(previous_loc_method)->method_string,
+				(char *)loc_method_api_get(requested_loc_method)->method_string);
+
+			loc_core_current_event_data_init(requested_loc_method);
+			err = loc_method_api_get(requested_loc_method)->location_get(
+				&current_loc_config.methods[current_loc_method_index]);
+			return;
+		}
+		LOG_ERR("Location acquisition failed and fallbacks are also done");
+	}
+
+	event_handler(&current_event_data);
+
+	if (current_loc_config.interval > 0) {
+		k_work_schedule_for_queue(
+			loc_core_work_queue_get(),
+			&loc_periodic_work,
+			K_SECONDS(current_loc_config.interval));
+	} else {
+		memset(current_loc_config.methods, 0,
+		       sizeof(struct loc_method_config) * LOC_MAX_METHODS);
+
+		k_sem_give(&loc_core_sem);
+	}
+}
+
+struct k_work_q *loc_core_work_queue_get(void)
+{
+	return &loc_core_work_q;
+}
+
+static void loc_core_periodic_work_fn(struct k_work *work)
+{
+	ARG_UNUSED(work);
+	loc_core_location_get(&current_loc_config);
+}
+
+static void loc_core_timeout_work_fn(struct k_work *work)
+{
+	enum loc_method current_loc_method =
+		current_loc_config.methods[current_loc_method_index].method;
+
+	ARG_UNUSED(work);
+
+	LOG_WRN("Timeout occurred");
+
+	loc_method_api_get(current_loc_method)->cancel();
+	loc_core_event_cb_timeout();
+}
+
+void loc_core_timer_start(uint16_t timeout)
```

Comment:
So timeout 0 == no timeout? Is this documented anywhere?

Reply:
We don't want immediate timeouts here because timeout default is 0 and I don't want `timeout > 0` checks to all methods when they start timer.
I will add a comment about this to location.h